src/vsr/replica_test.zig

```diff
@@ -1240,6 +1247,108 @@ test "Cluster: prepare beyond checkpoint trigger" {
     try expectEqual(t.replica(.R_).op_head(), checkpoint_1_prepare_max - 1);
 }
 
+test "Cluster: upgrade: operation=upgrade near trigger-minus-bar" {
+    const trigger_for_checkpoint = vsr.Checkpoint.trigger_for_checkpoint;
+    for ([_]struct {
+        request: u64,
+        checkpoint: u64,
+    }{
+        .{
+            // The entire last bar before the operation is free for operation=upgrade's, so when we
+            // hit the checkpoint trigger we can immediately upgrade the cluster.
+            .request = checkpoint_1_trigger - constants.lsm_batch_multiple,
+            .checkpoint = checkpoint_1,
+        },
+        .{
+            // Since there is a non-upgrade request in the last bar, the replica cannot upgrade
+            // during checkpoint_1 and must pad ahead to the next checkpoint.
+            .request = checkpoint_1_trigger - constants.lsm_batch_multiple + 1,
+            .checkpoint = checkpoint_2,
+        },
+    }) |data| {
+        const t = try TestContext.init(.{ .replica_count = 3 });
+        defer t.deinit();
+
+        var c = t.clients(0, t.cluster.clients.len);
+        try c.request(data.request, data.request);
+
+        t.replica(.R_).stop();
+        try t.replica(.R_).open_upgrade(&[_]u16{ 1, 2 });
+
+        // Prevent the upgrade from committing so that we can verify that the replica is still
+        // running version 1.
+        t.replica(.R_).drop(.__, .bidirectional, .prepare_ok);
+        t.run();
+        t.run();
+        try expectEqual(t.replica(.R_).op_checkpoint(), 0);
+        try expectEqual(t.replica(.R_).release(), 1);
+
+        t.replica(.R_).pass(.__, .bidirectional, .prepare_ok);
+        t.run();
+        t.run();
+        t.run();
```

Comment:
Do we need all three for the release to propagate?

Reply:
Ah, oops! It used to need those b/c I originally set the `upgrade_timeout` much higher.